### ISSUE TITLE
Add ASI (automatic semicolon insertion) support to semicolon rule

### ISSUE THREAD

**Issue by antoinerousseau** [palantir/tslint#363]
As explained in https://github.com/Microsoft/TypeScript/issues/2575

Some coders would like to forget about trailing semicolons, but in some cases it can be a source of errors that are undetected by both the TypeScript compiler and `tslint`.

Example 1:
```ts
function dangerous(): string {
    return
        "watch the semi-colons!"
}
dangerous()
```
Compiles into:
```js
function dangerous() {
    return;
    "watch the semi-colons!";
}
dangerous();
```

Example 2:
```ts
var foo = true
// some comment
['a', 'b'].forEach((l: string) => {/**/})
```
compiles into
```js
var foo = true['a', 'b'].forEach(function (l) {/**/})
```

That would be awesome to see some warnings in these cases, e.g. by turning on a new rule :heart_eyes:

**Comment by ashwinr** [comment]
@antoinerousseau the `semicolon` rule already does this for you :)

**Comment by antoinerousseau** [comment]
The idea was to allow coding without semicolons (and make it safe), not forcing them.

**Comment by ashwinr** [comment]
I see, so you want a "smarter" option that only flags a warning if it actually causes any harm?

**Comment by antoinerousseau** [comment]
Exactly! See discussion @ https://github.com/Microsoft/TypeScript/issues/2575

**Comment by ashwinr** [comment]
cool, i'm leaving this open for further discussion. not sure when we can get to this, but we'll consider it. thanks!

**Comment by thedmi** [comment]
+1

**Comment by adidahiya** [comment]
I think we should support this by allowing 2 options in the semicolon rule: `"always"` and `"never"`, just like ESLint: http://eslint.org/docs/rules/semi

**Comment by antoinerousseau** [comment]
`"always"` and `"never"` would be perfect indeed!

**Comment by WorldMaker** [comment]
I've been contemplating emulating [JavaScript Standard Style](http://standardjs.com/) in TSLint, and this would be a step in that direction.

**Comment by eyalsk** [comment]
I'd really love to see an option for "always" and "never" where never will follow ASI so in cases it isn't required it will complain about it.

At the moment I'm using semicolons because TSLint can only enforce it and I want to have a consistent code.

**Comment by Pajn** [comment]
I thought I should create a PR for this, but the check the special cases I would need to know what the first token in the next node is but I can't seem to find a good way to do that.

Anyone that can give a hint?

**Comment by adidahiya** [comment]
@Pajn I think you'll want to create a scanner to read the next line. You can check out some [existing code that uses `ts.Scanner`](https://github.com/palantir/tslint/search?utf8=%E2%9C%93&q=ts.Scanner).

**Comment by Pajn** [comment]
Thanks!

**Comment by JakeGinnivan** [comment]
Hey all,

As far as I can tell the `Example 2` is not fixed..

```
var foo = true
// some comment
['a', 'b'].forEach((l: string) => {/**/})
```

Is still dangerous and is not flagged when using semicolons never.

**Comment by adidahiya** [comment]
@JakeGinnivan can you please file that as a separate issue if it still exists in the latest version?
